In [2]:
import pandas as pd

In [81]:
df = pd.read_csv("Sample Data.csv")

In [36]:
df.head(20)

,Staff Key,faculty,school,broad_category,broad_contract_type,contract_type,contract_category,grade,gender,calc_fte,start_date,reporting_date
0,1140547C4,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,AC4,F,0.2,01/07/1998 00:00,31/05/2023 00:00
1,1140547C4,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,AC4,F,0.8,01/07/1998 00:00,31/05/2023 00:00
2,1140403 1,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,Professorial Range 1,M,0.4,01/08/2005 00:00,31/05/2023 00:00
3,1140403 1,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,Professorial Range 1,M,0.6,01/08/2005 00:00,31/05/2023 00:00
4,1140481 1,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,Professorial Range 1,M,0.2,01/07/2013 00:00,31/05/2023 00:00
5,1140481 1,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,Professorial Range 1,M,0.8,01/07/2013 00:00,31/05/2023 00:00
6,1100643 2,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,Professorial Range 2,F,0.3,27/03/2000 00:00,31/05/2023 00:00
7,1100643 2,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,Professorial Range 2,F,0.7,27/03/2000 00:00,31/05/2023 00:00
8,1100891 2,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,Professorial Range 2,M,0.3,01/07/2013 00:00,31/05/2023 00:00
9,1100891 2,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,Professorial Range 2,M,0.7,01/07/2013 00:00,31/05/2023 00:00


In [ ]:
# How many rows?
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Staff Key            30000 non-null  object 
 1   faculty              30000 non-null  object 
 2   school               30000 non-null  object 
 3   broad_category       30000 non-null  object 
 4   broad_contract_type  30000 non-null  object 
 5   contract_type        30000 non-null  object 
 6   contract_category    30000 non-null  object 
 7   grade                30000 non-null  object 
 8   gender               30000 non-null  object 
 9   calc_fte             30000 non-null  float64
 10  start_date           30000 non-null  object 
 11  reporting_date       30000 non-null  object 
dtypes: float64(1), object(11)
memory usage: 2.7+ MB


30000 rows. All data are non-null

In [42]:
# Dictionary of unique column values
unique_dict = {col: df[col].unique() for col in df.columns}
unique_dict

{'Staff Key': array(['1140547C4', '1140403 1', '1140481 1', ..., '1110482 7',
        '1100568 7', '1130583 6'], shape=(1336,), dtype=object),
 'faculty': array(['Arts, Humanities and Social Sciences', 'Belfast Region City Deal',
        'Corporate', 'Engineering and Physical Sciences'], dtype=object),
 'school': array(['Faculty Office AHSS',
        'Inst for Global Peace, Security and Justice',
        'Inst of Professional Legal Studies', "Queen's Business School",
        'School of Arts, English and Languages',
        'School of History, Anthropology, Philosophy and Politics',
        'School of Law',
        'School of Social Sciences, Education and Social Work', 'AMIC',
        'BRCD Directorate Office', 'iREACH Health', 'Momentum 1.0',
        'University General', 'Faculty Office EPS'], dtype=object),
 'broad_category': array(['Academic', 'Professional and Support', 'Academic Education',
        'Research'], dtype=object),
 'broad_contract_type': array(['Permanent', 'Fixed'],

In [30]:
# Do contract categories always make sense for the broad category they are associated with?
df.groupby('contract_category')['broad_category'].unique()

contract_category
Acad-Rel Administrative              [Professional and Support]
Acad-Rel Computer                    [Professional and Support]
Acad-Rel ORS                               [Academic Education]
Academic Non-Clinical                                [Academic]
Academic Non-Clinical (Education)          [Academic Education]
Catering/Ancillary/Sport             [Professional and Support]
Cleaning                             [Professional and Support]
Clerical                             [Professional and Support]
Clerical Intern/Student              [Professional and Support]
Operational Intern/Student           [Professional and Support]
Research                                             [Research]
Research Academic                                    [Research]
Teaching Cons/Asst/Assoc                   [Academic Education]
Technical                            [Professional and Support]
Technical Intern/Student             [Professional and Support]
Name: broad_category, 

In [31]:
df.groupby('contract_type')['broad_contract_type'].unique()

contract_type
Fixed-term                         [Fixed]
Fixed-term - SPC                   [Fixed]
Fixed-term - acting up             [Fixed]
Perm - acting up               [Permanent]
Perm on temp appt              [Permanent]
Permanent                      [Permanent]
Permanent (Flex Retirement)    [Permanent]
Name: broad_contract_type, dtype: object

All contract types belong to only one broad contract type. All contract categories belong to only one broad category. This looks healthy

Are there any duplicate rows?

In [82]:
df.duplicated().sum()

np.int64(142)

In [83]:
# Drop 142 duplicate rows
df = df.drop_duplicates()

In [84]:
# Identify employees with more than one contract
df_multiple_contracts = df[df["Staff Key"].duplicated(keep=False)]
df_multiple_contracts

,Staff Key,faculty,school,broad_category,broad_contract_type,contract_type,contract_category,grade,gender,calc_fte,start_date,reporting_date
0,1140547C4,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,AC4,F,0.2,01/07/1998 00:00,31/05/2023 00:00
1,1140547C4,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,AC4,F,0.8,01/07/1998 00:00,31/05/2023 00:00
2,1140403 1,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,Professorial Range 1,M,0.4,01/08/2005 00:00,31/05/2023 00:00
3,1140403 1,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,Professorial Range 1,M,0.6,01/08/2005 00:00,31/05/2023 00:00
4,1140481 1,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,Professorial Range 1,M,0.2,01/07/2013 00:00,31/05/2023 00:00
...,...,...,...,...,...,...,...,...,...,...,...,...
29995,1110535 6,Engineering and Physical Sciences,Faculty Office EPS,Professional and Support,Permanent,Permanent,Acad-Rel Administrative,Grade 6,F,1.0,27/08/2018 00:00,30/09/2023 00:00
29996,1110539 6,Engineering and Physical Sciences,Faculty Office EPS,Professional and Support,Permanent,Permanent,Acad-Rel Administrative,Grade 6,M,1.0,21/11/2018 00:00,30/09/2023 00:00
29997,1110532 7,Engineering and Physical Sciences,Faculty Office EPS,Professional and Support,Permanent,Permanent,Acad-Rel Administrative,Grade 7,M,1.0,14/05/2018 00:00,30/09/2023 00:00
29998,1110534 7,Engineering and Physical Sciences,Faculty Office EPS,Professional and Support,Permanent,Permanent,Acad-Rel Administrative,Grade 7,F,1.0,01/08/2018 00:00,30/09/2023 00:00


In [61]:
df_multiple_contracts[df_multiple_contracts["Staff Key"] == "1110547 7"]

,Staff Key,faculty,school,broad_category,broad_contract_type,contract_type,contract_category,grade,gender,calc_fte,start_date,reporting_date
29813,1110547 7,Engineering and Physical Sciences,Faculty Office EPS,Professional and Support,Permanent,Permanent,Acad-Rel Administrative,Grade 7,F,1.0,05/08/2019 00:00,31/05/2023 00:00
29858,1110547 7,Engineering and Physical Sciences,Faculty Office EPS,Professional and Support,Permanent,Permanent,Acad-Rel Administrative,Grade 7,F,1.0,05/08/2019 00:00,30/06/2023 00:00
29904,1110547 7,Engineering and Physical Sciences,Faculty Office EPS,Professional and Support,Permanent,Permanent,Acad-Rel Administrative,Grade 7,F,1.0,05/08/2019 00:00,31/07/2023 00:00
29951,1110547 7,Engineering and Physical Sciences,Faculty Office EPS,Professional and Support,Permanent,Permanent,Acad-Rel Administrative,Grade 7,F,1.0,05/08/2019 00:00,31/08/2023 00:00
29999,1110547 7,Engineering and Physical Sciences,Faculty Office EPS,Professional and Support,Permanent,Permanent,Acad-Rel Administrative,Grade 7,F,1.0,05/08/2019 00:00,30/09/2023 00:00


In addition to 142 duplicate rows containing the same information in each column, there also seems to be instances where the same contract has been reported on different dates. As these are not distinct contracts the duplicates can be removed

In [85]:
cols = df.columns.drop("reporting_date")
df = df.drop_duplicates(subset=cols)

In [86]:
df

,Staff Key,faculty,school,broad_category,broad_contract_type,contract_type,contract_category,grade,gender,calc_fte,start_date,reporting_date
0,1140547C4,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,AC4,F,0.2,01/07/1998 00:00,31/05/2023 00:00
1,1140547C4,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,AC4,F,0.8,01/07/1998 00:00,31/05/2023 00:00
2,1140403 1,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,Professorial Range 1,M,0.4,01/08/2005 00:00,31/05/2023 00:00
3,1140403 1,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,Professorial Range 1,M,0.6,01/08/2005 00:00,31/05/2023 00:00
4,1140481 1,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,Professorial Range 1,M,0.2,01/07/2013 00:00,31/05/2023 00:00
...,...,...,...,...,...,...,...,...,...,...,...,...
29839,1100535 5,Engineering and Physical Sciences,Faculty Office EPS,Professional and Support,Permanent,Perm on temp appt,Clerical,Grade 5,F,1.0,13/08/2018 00:00,30/06/2023 00:00
29841,1110482 7,Engineering and Physical Sciences,Faculty Office EPS,Professional and Support,Permanent,Perm on temp appt,Acad-Rel Administrative,Grade 7,F,1.0,19/07/2013 00:00,30/06/2023 00:00
29859,1100568 7,Engineering and Physical Sciences,Faculty Office EPS,Professional and Support,Permanent,Permanent,Acad-Rel Administrative,Grade 7,F,1.0,05/10/2021 00:00,30/06/2023 00:00
29930,1130583 6,Engineering and Physical Sciences,Faculty Office EPS,Professional and Support,Fixed,Fixed-term,Acad-Rel Administrative,Grade 6,F,1.0,28/08/2023 00:00,31/08/2023 00:00


There appears to be 1768 unique contracts. Much lower than the 30000 originally recorded. 

In [87]:
# Redefine dataframe of employees with multiple contracts
df_multiple_contracts = df[df["Staff Key"].duplicated(keep=False)]
df_multiple_contracts

,Staff Key,faculty,school,broad_category,broad_contract_type,contract_type,contract_category,grade,gender,calc_fte,start_date,reporting_date
0,1140547C4,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,AC4,F,0.2,01/07/1998 00:00,31/05/2023 00:00
1,1140547C4,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,AC4,F,0.8,01/07/1998 00:00,31/05/2023 00:00
2,1140403 1,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,Professorial Range 1,M,0.4,01/08/2005 00:00,31/05/2023 00:00
3,1140403 1,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,Professorial Range 1,M,0.6,01/08/2005 00:00,31/05/2023 00:00
4,1140481 1,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,Professorial Range 1,M,0.2,01/07/2013 00:00,31/05/2023 00:00
...,...,...,...,...,...,...,...,...,...,...,...,...
29829,1000489 3,Engineering and Physical Sciences,Faculty Office EPS,Professional and Support,Permanent,Permanent,Clerical,Grade 3,F,0.8,05/06/2017 00:00,31/05/2023 00:00
29831,1110533 7,Engineering and Physical Sciences,Faculty Office EPS,Academic Education,Permanent,Permanent,Acad-Rel ORS,Grade 7,M,1.0,16/07/2018 00:00,31/05/2023 00:00
29832,1002562 8,Engineering and Physical Sciences,Faculty Office EPS,Professional and Support,Permanent,Permanent,Acad-Rel Computer,Grade 8,M,1.0,01/09/1988 00:00,31/05/2023 00:00
29834,1020725 9,Engineering and Physical Sciences,Faculty Office EPS,Academic Education,Permanent,Permanent,Acad-Rel ORS,Grade 9,M,1.0,02/04/2001 00:00,31/05/2023 00:00


There appear to be 705 employees who have held more than one type of contract. How many employees have had contracts with more than one grade?

In [88]:
df_changing_grades = df_multiple_contracts[df_multiple_contracts.groupby("Staff Key")["grade"].transform("nunique") > 1]

In [89]:
df_changing_grades

,Staff Key,faculty,school,broad_category,broad_contract_type,contract_type,contract_category,grade,gender,calc_fte,start_date,reporting_date
145,1150582 1,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Professional and Support,Fixed,Fixed-term - SPC,Clerical Intern/Student,Grade 1,F,1.0,01/08/2023 00:00,31/08/2023 00:00
146,1150582 1,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Professional and Support,Fixed,Fixed-term - SPC,Clerical Intern/Student,Grade 1,M,1.0,01/08/2023 00:00,31/08/2023 00:00
1535,1150583 1,"Arts, Humanities and Social Sciences","Inst for Global Peace, Security and Justice",Academic,Permanent,Permanent,Academic Non-Clinical,Professorial Range 1,F,1.0,01/01/2024 00:00,31/01/2024 00:00
7292,1160596 1,"Arts, Humanities and Social Sciences",Queen's Business School,Professional and Support,Fixed,Fixed-term - SPC,Clerical Intern/Student,Grade 1,M,1.0,23/07/2025 00:00,31/07/2025 00:00
21308,1160599 2,"Arts, Humanities and Social Sciences",School of Law,Academic,Permanent,Permanent,Academic Non-Clinical,Professorial Range 2,F,1.0,01/04/2026 00:00,30/04/2026 00:00
27039,1150583 1,Belfast Region City Deal,AMIC,Professional and Support,Fixed,Fixed-term - SPC,Operational Intern/Student,Grade 1,M,1.0,15/08/2023 00:00,31/08/2024 00:00
27827,1160599 2,Belfast Region City Deal,AMIC,Professional and Support,Fixed,Fixed-term,Technical,Grade 2,M,1.0,05/11/2025 00:00,30/11/2025 00:00
27828,1160599 2,Belfast Region City Deal,AMIC,Professional and Support,Fixed,Fixed-term,Technical,Grade 2,M,1.0,19/11/2025 00:00,30/11/2025 00:00
28632,1150582 1,Belfast Region City Deal,Momentum 1.0,Research,Fixed,Fixed-term,Research,ECIT Grade 1,M,1.0,24/07/2023 00:00,31/08/2024 00:00
29054,1160596 1,Belfast Region City Deal,Momentum 1.0,Research,Permanent,Permanent,Research,ECIT Grade 1,F,1.0,16/07/2025 00:00,31/07/2025 00:00


In [90]:
df_changing_grades["Staff Key"].unique()

array(['1150582 1', '1150583 1', '1160596 1', '1160599 2', '1160598 1'],
      dtype=object)

Only 5 out of 1336 staff appear to have changed grades. There is an issue with the Staff Key whereby the grade is used at the end of the Staff ID number. This means Staff Key is not a unique identifier for a staff member, but rather for a contract. 

Let's go back and identify unique staff members

In [92]:
df_multiple_contracts["Staff ID"] = df_multiple_contracts["Staff Key"].str[:7]

C:\Users\Niall.Russell\AppData\Local\Temp\ipykernel_31684\2770146454.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_multiple_contracts["Staff ID"] = df_multiple_contracts["Staff Key"].str[:7]


In [93]:
df_multiple_contracts

,Staff Key,faculty,school,broad_category,broad_contract_type,contract_type,contract_category,grade,gender,calc_fte,start_date,reporting_date,Staff ID
0,1140547C4,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,AC4,F,0.2,01/07/1998 00:00,31/05/2023 00:00,1140547
1,1140547C4,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,AC4,F,0.8,01/07/1998 00:00,31/05/2023 00:00,1140547
2,1140403 1,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,Professorial Range 1,M,0.4,01/08/2005 00:00,31/05/2023 00:00,1140403
3,1140403 1,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,Professorial Range 1,M,0.6,01/08/2005 00:00,31/05/2023 00:00,1140403
4,1140481 1,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Academic,Permanent,Perm on temp appt,Academic Non-Clinical,Professorial Range 1,M,0.2,01/07/2013 00:00,31/05/2023 00:00,1140481
...,...,...,...,...,...,...,...,...,...,...,...,...,...
29829,1000489 3,Engineering and Physical Sciences,Faculty Office EPS,Professional and Support,Permanent,Permanent,Clerical,Grade 3,F,0.8,05/06/2017 00:00,31/05/2023 00:00,1000489
29831,1110533 7,Engineering and Physical Sciences,Faculty Office EPS,Academic Education,Permanent,Permanent,Acad-Rel ORS,Grade 7,M,1.0,16/07/2018 00:00,31/05/2023 00:00,1110533
29832,1002562 8,Engineering and Physical Sciences,Faculty Office EPS,Professional and Support,Permanent,Permanent,Acad-Rel Computer,Grade 8,M,1.0,01/09/1988 00:00,31/05/2023 00:00,1002562
29834,1020725 9,Engineering and Physical Sciences,Faculty Office EPS,Academic Education,Permanent,Permanent,Acad-Rel ORS,Grade 9,M,1.0,02/04/2001 00:00,31/05/2023 00:00,1020725


In [94]:
df_changing_grades = df_multiple_contracts[df_multiple_contracts.groupby("Staff ID")["grade"].transform("nunique") > 1]

In [95]:
df_changing_grades

,Staff Key,faculty,school,broad_category,broad_contract_type,contract_type,contract_category,grade,gender,calc_fte,start_date,reporting_date,Staff ID
37,1120571 7,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Professional and Support,Permanent,Permanent,Acad-Rel Administrative,Grade 7,F,1.0,07/02/2022 00:00,31/05/2023 00:00,1120571
145,1150582 1,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Professional and Support,Fixed,Fixed-term - SPC,Clerical Intern/Student,Grade 1,F,1.0,01/08/2023 00:00,31/08/2023 00:00,1150582
146,1150582 1,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Professional and Support,Fixed,Fixed-term - SPC,Clerical Intern/Student,Grade 1,M,1.0,01/08/2023 00:00,31/08/2023 00:00,1150582
239,1150584 2,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Professional and Support,Fixed,Fixed-term - SPC,Clerical,Grade 2,F,1.0,16/10/2023 00:00,31/10/2023 00:00,1150584
655,1120571 7,"Arts, Humanities and Social Sciences",Faculty Office AHSS,Professional and Support,Permanent,Permanent,Acad-Rel Administrative,Grade 7,F,0.6,07/02/2022 00:00,30/06/2024 00:00,1120571
...,...,...,...,...,...,...,...,...,...,...,...,...,...
29470,1160600 1,Belfast Region City Deal,Momentum 1.0,Research,Permanent,Permanent,Research,ECIT Grade 1,M,1.0,01/04/2026 00:00,30/04/2026 00:00,1160600
29820,1120556 7,Engineering and Physical Sciences,Faculty Office EPS,Academic Education,Fixed,Fixed-term,Acad-Rel ORS,Grade 7,M,1.0,25/05/2020 00:00,31/05/2023 00:00,1120556
29822,1130565 7,Engineering and Physical Sciences,Faculty Office EPS,Professional and Support,Fixed,Fixed-term,Acad-Rel Computer,Grade 7,M,1.0,17/05/2021 00:00,31/05/2023 00:00,1130565
29824,1120556 8,Engineering and Physical Sciences,Faculty Office EPS,Academic Education,Fixed,Fixed-term,Acad-Rel ORS,Grade 8,M,1.0,01/04/2020 00:00,31/05/2023 00:00,1120556
